### PR TITLE
[aie] Remove boost dependency by implementing Dijkstra using ADT

### PR DESCRIPTION
**Pull request by makslevental**
This PR excises the dependency on boost by using `llvm/ADT/DirectedGraph.h` types and implementing Dijkstra by hand. 

Other general cleanup is done in-place.

Currently, the only remaining issue is this implementation isn't deterministic in the ordering of the switchboxes and flows created; the routings it finds are still correct, they're just out of order  (hence the sprinkling of `CHECK-DAG`). Obviously it's preferrable that everything be deterministic so I'll keep hunting.

### TODO 

1. ~~Ablate the `Switchbox` and `Channel` constructor panoply~~
2. ~~use `shared_ptr` instead of `new Channel`~~
3. ~~fix `CHECK-DAG`~~
4. ~~add error throwing mechanism for max iters reached~~
5. ~~turn `SwitchSetting`, `PathEndPoint`, and `Flow` into structs.~~ (next refactor...)
6. ~~handle no matching channel in `addFixedConnection` correctly instead of silently.~~
7. ~~rename `AIEPathfinder.h` and `AIEPathfinder.cpp`.~~

### CODE REVIEW

**Comment by keryell** [comment]
It is a trade-off between reimplementing graph algorithms and using libraries.
Boost is pretty common, even if the BGL is old and should be replaced by https://github.com/pnnl/NWGraph nowadays.

**Comment by makslevental** [comment]
> It is a trade-off between reimplementing graph algorithms and using libraries. Boost is pretty common, even if the BGL is old and should be replaced by https://github.com/pnnl/NWGraph nowadays.

Depending on boost (or something else) for dijkstra's is like killing a fly with a nuclear bomb.

File: include/aie/Dialect/AIE/Transforms/AIEPathfinder.h [hunk 1]
```diff
@@ -76,34 +89,82 @@ typedef std::pair<Switchbox *, Port> PathEndPoint;
...
+  using nodes_iterator = SwitchboxGraph::iterator;
+  static NodeRef getEntryNode(SwitchboxGraph *DG) { return *DG->begin(); }
+  static nodes_iterator nodes_begin(SwitchboxGraph *DG) { return DG->begin(); }
+  static nodes_iterator nodes_end(SwitchboxGraph *DG) { return DG->end(); }
+};
+
+inline raw_ostream &operator<<(raw_ostream &OS, const Switchbox &S) {
+  OS << "Switchbox(" << S.col << ", " << S.row << ")";
+  return OS;
+}
```
Comment (jsetoain) [inline include/aie/Dialect/AIE/Transforms/AIEPathfinder.h:161]: Nice! :-)

**Review by jsetoain — APPROVED** [review]
Very nice clean-up!

**Comment by keryell** [comment]
> > It is a trade-off between reimplementing graph algorithms and using libraries. Boost is pretty common, even if the BGL is old and should be replaced by [pnnl/NWGraph](https://github.com/pnnl/NWGraph) nowadays.
> 
> Depending on boost (or something else) for dijkstra's is like killing a fly with a nuclear bomb.

Some exaggeration perhaps? ;-) I am surprised you are not using MLIR for this.
At the end, why using NIH industry-grade software while we can just reimplement everything with some potential bugs?

**Comment by makslevental** [comment]
> Some exaggeration perhaps?

Where's the exaggeration? We pull in boost for a function that is implemented in ~15 lines of source. In CI _all of boost is pulled in_ (a binary for current CI but in other places, such as `manylinux` containers, or Mac, [it has to be compiled from source](https://github.com/Xilinx/mlir-aie/actions/runs/6660006484/job/18100287523#step:7:574)[^1]). Note, the code around the function is not appreciably different - munging to and from boost data structures takes roughly the same amount of code as munging to and from llvm/ADT data structures. 

> ;-) I am surprised you are not using MLIR for this.
> At the end, why using NIH industry-grade software while we can just reimplement everything with some potential bugs?

I am using MLIR for this - I am using llvm/ADT data structures which are comparably bug-free. Note, again, the code around Dijkstra is bug-ridden/bug-free as function of the munging rather than the implementation of Dijkstra itself, which should be quite easily verified for correctness by all maintainers of this repo.

[^1]: Sure there are flags that can be passed to trim the build.

**Comment by stephenneuendorffer** [comment]
> > ;-) I am surprised you are not using MLIR for this.
> > At the end, why using NIH industry-grade software while we can just reimplement everything with some potential bugs?
> 
> I am using MLIR for this - I am using llvm/ADT data structures which are comparably bug-free. Note, again, the code around Dijkstra is bug-ridden/bug-free as function of the munging rather than the implementation of Dijkstra itself, which should be quite easily verified for correctness by all maintainers of this repo.

I'm a little surprised LLVM doesn't already have a shortest path library?

**Comment by makslevental** [comment]
> I'm a little surprised LLVM doesn't already have a shortest path library?

:shrug: I grepped for Dijkstra and didn't find anything except one-off implementations ([such as the one I cribbed](https://github.com/llvm/llvm-project/blob/5675f44cebf602931e884595c1d488bcedf6b8f2/llvm/lib/Transforms/Utils/SampleProfileInference.cpp#L680)).

**Review by erwei-xilinx — APPROVED** [review]

**Pull request merged**